File: 1 Two Sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, int> m;
        for(int i = 0;i < nums.size();++i){
            if(m[nums[i]] && nums[i] * 2 == target){
                //由同個數字組成的情況
                return {m[nums[i]] - 1, i};
            }
            m[nums[i]] = i+1;
        }
        for(auto& i:m){
            if(i.second && m[target - i.first])//有n就找找target-n
                return {i.second - 1, m[target - i.first] - 1};
        }
        return {};//並不會用到
    }
};

File: 1 Two Sum.py
class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        dict = {}
        for i in range(len(nums)):
            if nums[i] in dict and target == nums[i] * 2:
                return [dict[nums[i]], i]
            else:
                dict[nums[i]] = i
        for i in dict:
            #leetcode的環境有點怪，字典會自行排序，其他環境可能錯誤
            #if target - i in dict and i != target - i:
            if target - i in dict:
                return [dict[i], dict[target - i]]
        '''
        #結構簡單，無額外空間，但效率極差
        for i in range(len(nums)-1, -1, -1):
            if target - nums[i] in nums and nums.index(target - nums[i]) != i:
                return [nums.index(target - nums[i]), i]
        '''